students/pkoleyni/Lesson05/mailroom_3.py

```diff
@@ -0,0 +1,186 @@
+
+def sort_key(key):
+    """
+
+    :param key: Key defines the item in the list to be used for sorting
+    :return:
+    """
+    return key[1]
+
+
+def report():
+    """
+
+    :return: This function created new list with the items required in the report, sort the list and print the report
+    """
+
+
+    donor_names = [donor for donor in donor_list_dic]
+    total_given = [sum(donation_list(donor)) for donor in donor_names]
+    num_of_gifts = [len(donation_list(donor)) for donor in donor_names]
+    average_gifts = [t/g for t in total_given for g in num_of_gifts]
+
+    report_list = []
+    for i in range(len(donor_names)):
+        report_list.append([donor_names[i], total_given[i], num_of_gifts[i], average_gifts[i]])
+
+
+    sorted_report_list = sorted(report_list, key=sort_key, reverse=True)
+    report_head = ("{:<20}{:<20}{:<20}{:<20}".format('Donor Name', '| Total Given', '| Num Gifts ', '| Average Gift'))
+    print(report_head)
+    print('-' * (len(report_head)))
+    report_line= ('{:<21}{:<21}{:<21}{:<21.2f}'.format(item[0], item[1], item[2], item[3]) for item in sorted_report_list)
+    for i in report_line:
+        print (i)
+def send_letter_to_all():
+        write_letter_to_file('pooria koleyni')
+
+
+def file_name(name):
+    """
+
+    :param name: name of a donor
+    :return: string that will be be used as file_name for thank you letter file
+    """
+    d = datetime.datetime.today()
+    return name + '_{}_{}_{}'.format(d.year, d.month, d.day) + '.txt'
+
+
+def write_letter_to_file(name):
+    last_donation = donation_list(name)[-1]
+    line1 = 'Dear {},\n'.format(name)
+    line2 = '\tThank you for your very kind donation of ${:.2f}.\n'.format(last_donation)
+    line3 = '\tIt will be put to very good use.\n'
+    line4 = '\tSincerely,\n\t     -The Team'
+
+    file = file_name(name)
+    with open(file, 'w') as f:
+        f.write(line1.format(name))
+        f.write(line2.format(name))
+        f.write(line3.format(name))
+        f.write(line4.format(name))
+    print('Letter for "{}" is saved to "{}"'.format(name, file))
+
+
+def write_letter_for_everyone():
+    """
+    Get name of each donor from the donor_list_dic dictionary and call write_letter_to_file() function.
+    :return:
+    """
+    for donor in donor_list_dic.keys():
+        write_letter_to_file(donor)
+
+
+def quit():
+    sys.exit()
+
+
+def menu_selection():
+    """
+    Display Main menu, Using a Dictionary to switch
+    :return:
+    """
+    main_prompt = """
+    Select from the following options:\n
+    1. Send a Thank You to a single donor.
+    2. Create a report
+    3. Send letters to all donors.
+    4. Quit
+
+    """
+    main_menu_dic = {'1': thank_you,
+                     '2': report,
+                     '3': write_letter_for_everyone,
+                     '4': quit
+                     }
+    while True:
+        response = input(main_prompt)
+        if main_menu_dic[response]() == "exit menu":
+            break
```

Comment:
you don't have a function returning "exit menu" - you should remove that check.

Also, what user enters "5" - would you program crash? how would you address it (hint: what exception can you catch?)